### ISSUE TITLE
Implementing RadioMixin in View

### ISSUE THREAD

**Issue by scott-w** [marionettejs/backbone.marionette#2956]
I've been going through the code and realised `Marionette.View` doesn't use the `RadioMixin`. Is there any particular reason for this? Perhaps something to do with Views being more commonly created/destroyed?

**Comment by paulfalgout** [comment]
Yeah.. I'm not sure.. only one object can handle any particular request on a channel at one time.  _maybe_ it would be useful on views too.. but it could promote some bad patterns as well. 

For events it isn't hard to implement this on your own too:

**Comment by paulfalgout** [comment]
So after sitting on this a bit I do think this is best left as a a recipe rather than a feature...  radio requests can only have 1 replier and there are many instances where you have multiple copies of the same view.  I think this just opens us up to dealing with issues around people not understanding radio patterns.

**Comment by scott-w** [comment]
I think that's fair enough, it'll only end up confusing developers needlessly. We can put a ticket in for advanced documentation to show users how to do this.

**Comment by gdavis2** [comment]
Thanks for the followup @scott-w, and consideration @paulfalgout.
Upon further thought, a agree that it's best left as a recipe. I was just playing around with the 3.0 features and mistakenly assumed Mn.View extended Mn.Object. Interestingly, I have a case where I would like my view to respond to more than one radio channel so the standard pattern would have to be worked around if View acted like Object anyway.

**Comment by rafde** [comment]
How about allowing for this scenario
```
const MyView = Marionette.View.extend({
  radioEvents: {
    myChannel:{
      myChannelEvent: 'onMyChannelEvent',
      myChannelOtherEvent: 'onMyChannelOtherEvent',
    },
    mySpanishChannel:{
      mySpanishChannelEvent: 'onMySpanishChannelEvent',
      mySpanishChannelOtherEvent: 'onMySpanishChannelOtherEvent'
    },
  },
  onMyChannelEvent() { console.log('hey'); },
  onMyChannelOtherEvent() { console.log('hello'); },
  onMySpanishChannelEvent() { console.log('ey'); },
  onMySpanishChannelOtherEvent() { console.log('hola'); }
});
```

**Comment by bazineta** [comment]
bindEvents, rather than bindEntityEvents?

**Comment by rafde** [comment]
reply/request are one-to-one. So if you bind a request to a view, and you have another view of the same class, then the request is bound to the new view.

I don't think binding request is viable.

Having View's listen to channels is do-able since they are limited to one-to-one like reply/request

**Comment by bazineta** [comment]
I think it's a viable pattern for a private channel, i.e., rather than:

var channel = Radio.channel('foo')

Think of a view instance private channel, i.e.,

this.channel = new Radio.Channel()

Which would then be passed to children of the view via options.

Additionally, it's not out of the question to have a top-level view of which there would be only one, e.g, a dashboard view, with many children, so in such a case, even a global channel would be viable.

**Comment by rafde** [comment]
I am not sure what the scenario for having a private channel would help with, but is seems like an edge case.

All Backbone extending "Classes" are already private channels. `model.trigger('secret', 'foobar')`, view.triggerMethod('view secret')` and others. So you can play with that idea that is already available.

channels just help with being able to send events in order to skip implicit event bubbling to some other part of the View hierarchy that needs to be aware of some event.

**Comment by bazineta** [comment]
The issue in the case of the dashboard example is that I'm in general not interested in events that children generate; rather, they're interested in ones that I generate. That is, in this case the events aren't bubbling up; rather, they need to be sent down. For example, during a drag and drop operation, the dashboard needs to inform views below it to disable themselves, e.g., undelegating their events, for the duration of the drag and drop reorganization, then, informing them to reenable themselves when the reorganization is complete. The components that would require such notifications may be a long way down in a view hierarchy from the dashboard.

**Comment by bazineta** [comment]
I've spent some time exploring this pattern, and frankly, I think it's a good one; in our application there are many instances where a containing view is acting as a mediator for hierarchies of contained views, some quite deep in the hierarchy, and being able to define radioEvents and call
```
 this.bindEvents(channel, this.radioEvents); 
```
works well in our case. In this manner the subordinate views may subscribe to events issued by the mediator view; it really cleans up the situation we used to have, where some type of communication object was passed down the hierarchy.

However, it's not uncommon for the mediator view to also hold data of interest to the subordinate views, so having it handle requests is also useful.  The following isn't quite as nice:
```
Marionette.bindRequests(this, channel, this.radioRequests);
```
and then, typically in an onDestroy or onBeforeDetach:
```
Marionette.unbindRequests(this, channel, this, radioRequests);
```
If there's a better approach to this, I'd appreciate suggestions; in this case, the mediator view at the top of the hierarchy has a model that contains the information of interest to the channel requestors. Not all of them require data from this model, so being able to request it from the channel when needed is a nice approach.

It seems that being able to call
```
this.bindRequests(channel, this.radioRequests);
```
would be a nicer approach. I think it not uncommon that there exist singleton views -- navigation bars, the root view, things like that, so being able to bind requests in a less verbose manner would be helpful.

And note, this isn't bubbling; bubbles float the the surface. I suppose this would be 'sinking'; i.e., the events are going down a hierarchy, not up.

**Comment by paulfalgout** [comment]
My argument is still that this is better handled with documentation.

Our current Radio mixin supports a single channel for events and requests.  You set the `channelName` on the Object and so the expectations of functionality is very straight-forward.

The common use-case for views is different.  In _most_ cases you don't want requests and in many cases you may not want a single channel.

I don't think we want to introduce either a different radioEvents API than object nor do I think radioRequests should be something that can be done without consideration.

FWIW @bazineta you do not need to cleanup the requests in an `onDestroy`  `this.bindRequests` is just a proxy for `Marionette.bindRequest(this`  Notice that `unbindRequests` is not called at all within the library at any point.  This is because they are bound to the lifecycle of the object.. so a normal `destroy` will clean them up.

I don't think this is a ton of boilerplate to support what would be a custom API for this particular view:

```js
const MyView = Marionette.View.extend({
  initialize() {
    const myChannel = Radio.channel('myChannel');
    const mySpanishChannel = Radio.channel('mySpanishChannel');
    Marionette.bindRequests(this, myChannel, this.radioRequests);
    this.bindEvents(myChannel, this.radioEvents.myChannel);
    this.bindEvents(mySpanishChannel, this.radioEvents.mySpanishChannel);
  },
  radioRequests: {
    myRequest: 'getRequest'
  }
  radioEvents: {
    myChannel:{
      myChannelEvent: 'onMyChannelEvent',
      myChannelOtherEvent: 'onMyChannelOtherEvent'
    },
    mySpanishChannel:{
      mySpanishChannelEvent: 'onMySpanishChannelEvent',
      mySpanishChannelOtherEvent: 'onMySpanishChannelOtherEvent'
    }
  },
  onMyChannelEvent() { console.log('hey'); },
  onMyChannelOtherEvent() { console.log('hello'); },
  onMySpanishChannelEvent() { console.log('ey'); },
  onMySpanishChannelOtherEvent() { console.log('hola'); },
  getRequest() { return 1; }
});
```

What might be interesting.. though I don't know if it's necessary is to add a `bindRadioRequests` and `bindRadioEvents` where the first argument is just the channel name as a string or the channel.

**Comment by rafde** [comment]
> though I don't know if it's necessary is to add a bindRadioRequests and bindRadioEvents where the first argument is just the channel name as a string or the channel.

could be
```
radioEvents:{
  'channel event': 'onChannelEvent'
}
```

but I still don't think it'd work so obviously with Request because of the one-to-one scenario.

If you need to do something for a request maybe just

```
// replying view
onDrag(){
  Radio.channel('dragNdrop').reply('draggingView', view);
  Radio.channel('dragNdrop').trigger('dragIsComing', view);
}
```

```
// requesting view
onDragEnter(){
  const draggingView = Radio.channel('dragNdrop').request('draggingView');
  if(this.shouldDoSomething(draggingView)){
    // do something
  }
},
onDragIsComing(view){
  // do something for this event.
}
```

**Comment by paulfalgout** [comment]
I'd maintain that I don't think View's should get a `radioEvents` hash.  I don't think it should support a different API than Object, and I don't think Object should support more than one channel.

**Comment by bazineta** [comment]
Thanks for the instruction on not having to clean up via unbindRequests(); that was not obvious to me, and the fact that it was exposed made me err on the side of caution.

To me at least, bindEvents(), documented and without any supporting magic, is perfectly fine for use in a view; the boilerplate seems straightforward and there is advantage to being able to support many channels.

I think that if bindRequests() was also available on the view, it'd lend some consistency, as the 2-arg vs. 3-arg nature of the two calls caused some initial confusion on my part.

**Comment by paulfalgout** [comment]
I think I can get behind adding `bindRequests` to the View.  I've caught myself assuming it was there.  But I still think a `radioRequests` hash on a view is just asking for introducing problems.  This should be a clear, purposeful omission to guard against error.

**Comment by paulfalgout** [comment]
Closing for #3492 
It doesn't 100% solve the original issue, but I think it the most ideal solution.  If this needs further discussion, we can reopen.

**Comment by bazineta** [comment]
I have some leak question related to this in #3492; copying it here for context. Above, on 17 September, @paulfalgout seems to indicate that there's no need to call `unbindRequests`. However, this fiddle:

https://jsfiddle.net/orj1sckr/5/

Seems to indicate otherwise. There, I create a view and call `bindRequests`. I replace that view with a second, then a third. In the third, I make a request on the channel that the first view was listening on, and it, despite being destroyed, answers.

**Comment by paulfalgout** [comment]
I need to dig into this a bit more but at first glance, adding an unbindRequests, didn't seem to change anything... Hmmm

**Comment by bazineta** [comment]
Previously, in our application, we did add an unbindRequests (3-parameter version), typically in the view's onDestroy(), and that did clean things up. We had removed it per the discussion above, but happened to trace an entertaining issue to this today, thus the observation.

The trace shows unbindEvents() with respect to Radio being called as expected, but unbindRequests() is never called at all unless we do so ourselves.

**Comment by paulfalgout** [comment]
Meh.. was trying adding `unbindRequests` from my phone.. it does seem to work as normal.. ok investigating.. FWIW `unbindEvents` is never called with, yet events are not continually bound.  I'm not suggesting that `unbindRequests` is necessarily called.

**Comment by paulfalgout** [comment]
ok so `unbindRequests` does not work unless the 3-parameter version is used.  This is unfortunate and might relate to the situation we're seeing.  Previously I was not suggesting that `unbindEvents` was ever called, but like `unbindRequests` never being called it should still remove the attachments from the view.

**Comment by paulfalgout** [comment]
Ha. this is because only the 3-parameter version works with binds, unlike `stopListening` and the like.

**Comment by paulfalgout** [comment]
Re: "The trace shows unbindEvents() with respect to Radio being called as expected"  `unbindEvents` is only called prior to binding and not called during destroy

**Comment by paulfalgout** [comment]
so I definitely agree that cleaning up of requests will be necessary for views, but not for objects/application because of https://github.com/marionettejs/backbone.marionette/blob/70fb35b20ddb47f99329a61116e479f6299df087/src/mixins/radio.js#L50

There's really is no automated way to clean this up as we don't know what the channel is, it'll have to be documented

 but there's a few problems here.  `bindRequest` and `bindEvents` for that matter don't store the entity being bound nor do they store the handlers and this is a problem.  particularly since the handlers are from a `_.result` they can be dynamic.

but a simple example... if a user replaces a model on a view (which is less than ideal anyhow) they'll need to recall `bindEntityEvents` in order to get `modelEvents` to work again.
Fortunately `bindEntityEvents` calls `unbindEntityEvents` first.. so we're all good correct?

**Nope**

in this case and in many cases the unbind would be called on the currently planned bind arguments, and not on what was bound leading to leaks.  This is probably a separate issue, but I'm going to reopen this until it can be nailed down.. might even be a v3.x fix

**Comment by bazineta** [comment]
Re. "There's really is no automated way to clean this up as we don't know what the channel is, it'll have to be documented" and "unbindEvents is only called prior to binding and not called during destroy".

It would seem then that the proxied functions (`bindEvents` in 3.x, `bindEvents` and `bindRequests` in next) are just inherently unusable from a leak perspective, and the only safe approach is to call the 3-arg versions only, with the 3-arg unbind versions being called in, presumably, `onDestroy` or one of the other teardown notification functions.

Is that  a reasonable assessment?

**Comment by paulfalgout** [comment]
well `unbindEvents` _may_ but not necessarily be leaky for the life of the view, but will inevitably be cleaned up when the view is destroyed.  not the case for requests though.  the function version on the instances is fine, but the unbind must be given the same instance and handlers as the bind.  not something that everyone is necessarily going to account for.. ideally unbind can handle only unbinding any handlers from the context for the channel..  that would at least provide a safe method for an `onDestroy`  Then we should probably remove the unbinds that are called from the binds as they're misleadingly suggesting the methods are safe.

**Comment by bazineta** [comment]
Let me rephrase then, since I'm trying to get my leaks plugged for 3.x and make this simple to understand for others, presuming that this isn't something that's entirely straightforward to resolve. Let's consider two calls, both made from the constructor or `initialize`; we consider in this case a leak to be 'the event handler lives on *after* the view is destroyed':

This:
```
this.bindEvents(options.channel, this.radioEvents)
```
*Will not leak*, as there will be cleanup when the view is destroyed.

But this (using the current non-proxied 3.x API):
```
Marionette.bindRequests(this, options.channel, this.radioRequests)
```
*Will leak*; the request handlers will live until the channel is destroyed, which is in many cases never. The only way to avoid a leak is to call the corresponding 3-arg `unbindRequests` in `onDestroy`.

Do I have this correct?

**Comment by paulfalgout** [comment]
yes and options.channel and radioRequests must be identical at the time of the unbind

**Comment by paulfalgout** [comment]
I think the reason this issue was reopened was mostly resolved in v3.5
